lib/active_record/connection_adapters/sqlserver/jdbc/database.rb

```diff
@@ -0,0 +1,361 @@
+          def connection_info(conn)
+            "(conn: #{conn.__id__}) "
+          end
+
+          # Log message with message prefixed by duration at info level, or
+          # warn level if duration is greater than log_warn_duration.
+          def log_duration(duration, message)
+            log_each((lwd = log_warn_duration and duration >= lwd) ? :warn : sql_log_level, "(#{sprintf('%0.6fs', duration)}) #{message}")
+          end
+
+          # Log message at level (which should be :error, :warn, or :info)
+          # to all loggers.
+          def log_each(level, message)
+            @loggers.each{|logger| logger.send(level, message)}
+          end
+        end
+
+        # connection
+        class Database
+          # The Java database driver we are using (should be a Java class)
+          attr_reader :driver
+
+          # The fetch size to use for JDBC Statement objects created by this database.
+          # By default, this is nil so a fetch size is not set explicitly.
+          attr_accessor :fetch_size
+
+          attr_reader :opts
+
+          attr_reader :conn
+
+          # Connects to a database.
+          def self.connect(conn_string, opts = OPTS)
+            opts = {:uri=>conn_string}.merge!(opts)
+            # process opts a bit
+            opts = opts.inject({}) do |m, (k,v)|
+              k = :user if k.to_s == 'username'
+              m[k.to_sym] = v
+              m
+            end
+            db = self.new(opts)
+            db.connect(opts)
+            db
+          end
+
+          def initialize(opts)
+            @opts = opts
+            @loggers = Array(@opts[:logger]) + Array(@opts[:loggers])
+            @sql_log_level = @opts[:sql_log_level] ? @opts[:sql_log_level].to_sym : :info
+            @default_dataset = Dataset.new(self)
+            @driver = com.microsoft.sqlserver.jdbc.SQLServerDriver
+            @fetch_size = nil
+            self.database_timezone = opts[:database_timezone]
+            raise(Error, "No connection string specified") unless uri
+          end
+
+          def database_timezone
+            opts[:database_timezone]
+          end
+
+          def database_timezone=(zone)
+            zone = (zone == :local ? :local : :utc)
+            opts[:database_timezone] = zone
+          end
+
+          # Connect to the database using JavaSQL::DriverManager.getConnection.
+          def connect(opts)
+            conn = if jndi?
+              get_connection_from_jndi
+            else
+              args = [uri(opts)]
+              args.concat([opts[:user], opts[:password]]) if opts[:user] && opts[:password]
+              begin
+                JavaSQL::DriverManager.setLoginTimeout(opts[:login_timeout]) if opts[:login_timeout]
+                raise StandardError, "skipping regular connection" if opts[:jdbc_properties]
+                JavaSQL::DriverManager.getConnection(*args)
+              rescue JavaSQL::SQLException, *DATABASE_ERROR_CLASSES, StandardError => e
```

Comment:
Relates to my comment about `DATABASE_ERROR_CLASSES` is this rescuing a lot more than expected?